### PR TITLE
Detect proxy pac configuration

### LINKED ISSUE #196: Problems automatically detecting the proxy
Does not detect proxy if configured via automatic configuration scripts (pac)

### PR DESCRIPTION
**Pull request by jenriq**
Fix #196

### CODE REVIEW

**Comment by pb00068** [comment]
This is a interesting PR.
Question: it this designed to read registry on 32 bit systems only?
(This would by a pity since most systems nowadays are 64 bit).

**Comment by jenriq** [comment]
> 
> 
> This is a interesting PR.
> Question: it this designed to read registry on 32 bit systems only?
> (This would by a bit taint since most systems nowadays are 64 bit).

works only in 32 bits, the PR is based on the 32-bit dll that already included the project

File: launcher/src/main/java/com/threerings/getdown/launcher/ProxyUtil.java [hunk 1]
```diff
@@ -96,7 +123,28 @@ public static boolean autoDetectProxy (Application app)
         initProxy(app, host, port, null, null);
         return true;
     }
-
+    
+    
+    
+    private static String findProxyForURL(String upac, Object... args) {
+    	ScriptEngineManager manager = new ScriptEngineManager();
+    	ScriptEngine engine = manager.getEngineByName("javascript");
```
Comment (pb00068) [inline launcher/src/main/java/com/threerings/getdown/launcher/ProxyUtil.java]: Here using openjdk11 I only get back an not-null instance when using javascript with capitals: 
`ScriptEngine engine = manager.getEngineByName("JavaScript");`

File: launcher/src/main/resources/com/threerings/getdown/launcher/PacUtils.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	var bytes = ipchars.split('.');
+	var result = ((bytes[0] & 0xff) << 24) |
+		((bytes[1] & 0xff) << 16) |
+		((bytes[2] & 0xff) <<  8) |
+		(bytes[3] & 0xff);
+	return result;
+}
+
+function isInNet(ipaddr, pattern, maskstr){
+	var test = /^(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})$/(ipaddr);
```
Comment (pb00068) [inline launcher/src/main/resources/com/threerings/getdown/launcher/PacUtils.js]: This  line causes following exception in our environment:

javax.script.ScriptException: TypeError: /^(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})$/ is not a function in <eval> at line number 23
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.throwAsScriptException(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.invokeImpl(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.invokeFunction(Unknown Source)
	at com.threerings.getdown.launcher.ProxyUtil.findProxyForURL(ProxyUtil.java:187)
	at com.threerings.getdown.launcher.ProxyUtil.autoDetectProxy(ProxyUtil.java:133)
	at com.threerings.getdown.launcher.Getdown.detectProxy(Getdown.java:200)
	at com.threerings.getdown.launcher.Getdown.run(Getdown.java:143)
Caused by: <eval>:23 TypeError: /^(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})$/ is not a function
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ECMAErrors.error(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ECMAErrors.typeError(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ECMAErrors.typeError(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ECMAErrors.typeError(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptObject.notAFunction(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptObject.findCallMethod(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptObject.lookup(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(Unknown Source)
	at jdk.dynalink/jdk.dynalink.linker.support.CompositeTypeBasedGuardingDynamicLinker.getGuardedInvocation(Unknown Source)
	at jdk.dynalink/jdk.dynalink.linker.support.CompositeGuardingDynamicLinker.getGuardedInvocation(Unknown Source)
	at jdk.dynalink/jdk.dynalink.LinkerServicesImpl.lambda$getGuardedInvocation$0(Unknown Source)
	at jdk.dynalink/jdk.dynalink.LinkerServicesImpl.getWithLookupInternal(Unknown Source)
	at jdk.dynalink/jdk.dynalink.LinkerServicesImpl.getGuardedInvocation(Unknown Source)
	at jdk.dynalink/jdk.dynalink.DynamicLinker.relink(Unknown Source)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$10$531AAA$\^eval\_.isInNet(<eval>:23)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$4$36AA$\^eval\_.:scopeCall(<eval>)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$4$36AA$\^eval\_.FindProxyForURL(<eval>:40)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptFunction.invoke(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptRuntime.apply(Unknown Source)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.ScriptObjectMirror.callMember(Unknown Source)
	... 6 more

Comment (pb00068) [reply launcher/src/main/resources/com/threerings/getdown/launcher/PacUtils.js]: And that a call to function exec is missing:
var test = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/.exec(ipaddr);

**Comment by samskivert** [comment]
So, no offense intended, but this PR was kind of a mess. As far as I could tell, Getdown didn't even compile with the patch applied (at least Proguard failed when I tried to do a full build), and the code did not appear to work at all when I tested it.

When I fixed the obvious broken stuff, then we got to the subtler broken stuff, and the important things missing from PacUtil.js and so forth.

In any case, I think it's a worthwhile addition, in spite of my horror at the fact that one has to use an entire JavaScript VM to determine proxy configuration information. At least it's largely opt in. So I rewrote the PR and added tests to make sure the JavaScript part works.

I have not tested it against a real PAC proxy because I don't have access to one, but I hope that since the individual parts are tested, it all works correctly when the parts are combined together. I assume you'll be giving this a shot in a real environment since you wanted the PR in the first place, so let me know if you run into any issues.

There is one question unresolved in my commit, which is the fact that your PR ignored the `ProxyEnable` Windows registry entry when the `AutoConfigURL` entry was set. Was that on purpose or an oversight? Does `ProxyEnable` not apply to proxy settings obtained via `AutoConfigURL`? Or is it possible for someone to configure an `AutoConfigURL` proxy and then disable it, such that we would see `AutoConfigURL` but also see `ProxyEnable=false`. Vice versa, if it is possible for us to see `AutoConfigURL` but _not_ see a `ProxyEnable` entry, then we do need to implicitly enable proxy use when we see that config.

**Pull request closed without merging**